First version of informative README. (#1)
wasp-lang/wasp: README.md

# How does it work

Source files (.wasp, .js(x), .css, .html, ...) are compiled (transpiled) by `waspc` (Wasp compiler) into the web technology stack of your choice.

Generated code looks as if it was handwritten by a senior engineer, following industry best-practices.
This is only really important if you decide Wasp is not flexible enough for you, then you can take the generated code and continue working from it.
Otherwise, you don't need to know what generated code looks like.

# This repository

This is the main repo of the Wasp universe, containing core code (`waspc` code is coming soon) and supporting materials.

# Documentation

While we do plan to have detailed documentation, we don't have anything worth publishing yet due to how fast everything is changing in this early stage of Wasp development.

In the meantime, the best way to learn more is by taking a look at our [TodoApp example](examples/TodoApp).

# Project status

> TODO: Describe current progress, where we are, what is the plan. Concisely, in just a couple of sentences, as high-level overview. We want to mention that we are very early and are now doing closed development of , but that generation of that code really works, and that we plan to get waspc to level where it can serve as useful generator, and then get it from that to level where will people will be ejecting later and later. Say which tech stack are we using right now.

Author: I have to finish this one before merging!

```
                                                                 _  _
                             __      ____ _ ___ _ __            | )/ )
                             \ \ /\ / / _` / __| '_ \        \\ |//,' __
                              \ V  V / (_| \__ \ |_) |       (")(_)-"()))=-
                               \_/\_/ \__,_|___/ .__/           (\\
                                               |_|
```

<p align=center> A programming language that understands what a web app is. </p>
<br>

[![Gitter](https://badges.gitter.im/wasp-lang/community.svg)](https://gitter.im/wasp-lang/community?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge)

<br>

Wasp (**W**eb **A**pplication **Sp**ecification Language) is an extensible [DSL](https://en.wikipedia.org/wiki/Domain-specific_language) (domain-specific language) for building modern web apps with less code.

> This means that concepts such as *app*, *page*, *user*, *login*, *frontend*, *production* etc. are baked into the language, bringing a new level of expressiveness and allowing you to get a lot of work done with very few lines of code.

Reviewer: ```suggestion
This means that concepts such as *app*, *page*, *user*, *login*, *frontend*, *production* etc. are baked into the language, bringing a higher level of expressiveness and allowing you to get more work done with a fewer lines of code.
```

```js
// todoApp.wasp:

app TodoApp {
  title: "ToDo app",
  favicon: "./todo-logo.png"
}

entity Task {
  description :: string,
  isDone :: boolean
}

entity-list<Task> TaskList {

>   editable: true

Reviewer: this we don't have yet, but ok, can easily add it

}

page Main {
  route: "/",
  content: {=jsx <TaskList/> jsx=},
  style: "./main.css"
}

...
```

> Source files (.wasp, .js(x), .css, ...) are compiled (transpiled) by `waspc` (Wasp compiler) into the web technology stack of your choice (e.g. React + Redux + ExpressJS + ...).

Reviewer: ```suggestion
Source files (`.wasp`, .`js(x)`, `.css`, ...) are compiled (transpiled) by `waspc` (Wasp compiler) into the web technology stack of your choice (e.g. React + Redux + ExpressJS + ...).
```

Why is Wasp awesome:
- **Quick start**: Due to its expressiveness, you can create and deploy a production-ready web app from scratch with very few lines of concise, consistent, declarative code.
- **Flexible**: When you need more control than Wasp offers, you can write code in existing technologies such as js/html/css/... and combine it with Wasp code!
- **No lock-in**: If Wasp becomes too limiting for you, simply eject and continue with the generated source code that follows industry best-practices, as if it was handwritten by a senior engineer.

For more of general information about Wasp, check:
- **The main webpage**: [https://wasp-lang.dev].
- [**Blog**](https://blog.wasp-lang.dev/): [Why are we building Wasp?](https://blog.wasp-lang.dev/posts/2019-09-01-hello-wasp.html)

Check out [TodoApp example](examples/todoApp) for complete code example.

> If you are interested/excited about Wasp, **support us by giving us a star :star: and/or [subscribe](https://wasp-lang.dev#signup) to our email list**! Check [Contributing](#contributing) for more information.

Reviewer: ```suggestion
If you are interested in/excited about Wasp, **support us by giving us a star :star: and/or [subscribe](https://wasp-lang.dev#signup) to our email list**! Check [Contributing](#contributing) for more information.
```

# This repository

> This is the main repo of the Wasp universe, containing core code (not really at the moment, but `waspc` (Wasp compiler) code is coming soon) and supporting materials.

Reviewer: ```suggestion
This is the main repo of the Wasp universe, containing core code (not at the moment, but `waspc` (Wasp compiler) code is coming soon) and the supporting materials.
```

# Documentation

> While we do plan to have detailed documentation, we don't have anything worth publishing yet due to how fast everything is changing in this early stage of Wasp development.

Reviewer: ```suggestion
While we do plan to publish detailed documentation, we don't have anything worth publishing yet due to how fast everything is changing in this early stage of Wasp development.
```

In the meantime, the best way to learn more is by taking a look at our [TodoApp example](examples/todoApp).

# Project status

We are still early and are doing closed development iterations with our early adopters.

Specifically, we are developing `waspc` (Wasp compiler) and also designing language as we go.

Currently, `waspc` is in a state where the main parts are there (code analysis and generation) and we are successfully generating code as can be seen at [TodoApp example](examples/todoApp).

Language itself supports a narrow set of features for now, not enough for serious web app development, but it is good basis to build on, we believe.
The next step is to expand the language (and compiler) so it can serve the role of a powerful code generator - it will still not have enough features to keep you from ejecting for very long time, but it will be enough to give you a quick start.

> Finally, as it grows further, it should allow development of web apps of increasing complexity without the need to eject!

Reviewer: nice paragraph

> While the idea is to support multiple web tech stacks in the future, right now we are focusing on specific stack: React + Redux, NodeJS + ExpressJS, Mongo. We might yet change that as time goes, taking trends into account, but for now this is serving us well for the purpose of developing compiler and language.

Reviewer: ```suggestion
While the idea is to support multiple web tech stacks in the future, right now we are focusing on the specific stack: React + Redux, NodeJS + ExpressJS, Mongo. We might yet change that as time goes, taking trends into account, but for now this is serving us well for the purpose of developing compiler and language.
```

# Contributing

Since we are still doing closed development of very first version of compiler (`waspc`) and language (check [Project status](#project-status) for more info), you can't contribute to the code directly via this repository at the moment.

However, even more than help with development, at this early stage we need your support and feedback! You can help us greatly by:
 - :star: Starring this repository.
 - :mailbox: Staying updated by subscribing to our email list at [wasp-lang.dev](https://wasp-lang.dev#signup).
 - :speech_balloon: Providing feedback and ideas via gitter [![Gitter](https://badges.gitter.im/wasp-lang/community.svg)](https://gitter.im/wasp-lang/community?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge).

>    Also let us know if you would like to be more closely involved, as we are looking for developers to participate with us in designing / testing the closed version of Wasp before we open it.

Reviewer: ```suggestion
   Also let us know if you would like to be more closely involved, as we are looking for developers to participate with us in designing / testing the closed version of Wasp before we open it to the general public.
```
